### PR TITLE
Add API: Change password

### PR DESCRIPTION
**Pull request by houtaru**

### CODE REVIEW

File: controller/auth.js [hunk 1]
```diff
@@ -17,9 +17,14 @@ function isEmail(email_or_username) {
   return emailRegex.test(email_or_username);
 }
 
-function sanitizeUserDetails(data) {
+function getHashPassword(password) {
```
Comment (dothanhlam97) [inline controller/auth.js]: nit: getHashedPassword

File: controller/auth.js [hunk 2]
```diff
@@ -17,9 +17,14 @@ function isEmail(email_or_username) {
...
   const salt = bcrypt.genSaltSync(SALT_ROUNDS);
-  const hashedPassword = bcrypt.hashSync(data.password, salt);
+  const hashedPassword = bcrypt.hashSync(password, salt);
```
Comment (dothanhlam97) [inline controller/auth.js]:
```
const hashedPassword = bcrypt.hashSync(password, salt);
 return hashedPassword;
```
-> `return bcrypt.hashSync(password, salt);`

File: routes/index.js [hunk 3]
```diff
@@ -63,11 +64,24 @@ router.get(
   makeHandlerAwareOfAsyncErrors(isLoggedIn),
   makeHandlerAwareOfAsyncErrors(userController.getAll),
 );
+router.get(
```
Comment (dothanhlam97) [inline routes/index.js]: Có sẽ e merge code bị lỗi chỗ này. 2 API get users nè

File: routes/index.js [hunk 4]
```diff
@@ -33,6 +33,7 @@ function makeHandlerAwareOfAsyncErrors(handler) {
           data: error.data || {},
         });
       } else {
+        console.log(error);
```
Comment (dothanhlam97) [inline routes/index.js]: Remove console.log

File: controller/user.js [hunk 5]
```diff
@@ -198,6 +201,38 @@ async function update(req, res) {
...
+  if (!rowsChange[0]) {
+    throw new errors.FcError(errors.BAD_REQUEST);
+  }
+
+  const updatedUser = await models.User.findByPk(user_id);
+
+  res.status(statusCode.SUCCESS).json({
+    code: 0,
+    msg: "Password updated",
+    data: { },
```
Comment (dothanhlam97) [inline controller/user.js]: Nit: data: {}

File: controller/user.js [hunk 6]
```diff
@@ -198,6 +201,38 @@ async function update(req, res) {
...
+
+  const hashedPassword = getHashPassword(req.body.new_password);
+
+  const rowsChange = await models.User.update({
+    password: hashedPassword
+  }, {
+    where: { id: user_id },
+  });
+
+  if (!rowsChange[0]) {
```
Comment (dothanhlam97) [inline controller/user.js]: Y chang PR trước

File: controller/user.js [hunk 7]
```diff
@@ -5,13 +5,16 @@ const db = require("../models/index.js");
...
+const bcrypt = require("bcryptjs");
+const { getHashPassword } = require("./auth.js");
 
 function buildUserJson(user) {
   return {
     id: user.id,
     username: user.username,
     full_name: user.full_name,
     school_name: user.school_name,
+    email: user.email,
```
Comment (dothanhlam97) [inline controller/user.js]: Mình thống nhất không return email khi get users trừ get me  
Cho nên nếu e muốn return email, my suggestion is 
function buildUserJson(user, can_show_email=False)

Comment (kc97ble) [reply controller/user.js]: rename `buildUserJson` to `formatUser`

File: controller/user.js [hunk 8]
```diff
@@ -198,6 +201,38 @@ async function update(req, res) {
   });
 }
 
+async function changePassword(req, res) {
+  const user_id = !req.params.id ? req.user.id : getIdParam(req);
```
Comment (kc97ble) [inline controller/user.js]: ????

We no longer use `id`, please check the API docs.`

File: controller/auth.js [hunk 9]
```diff
@@ -178,12 +182,13 @@ async function signup(req, res) {
   res.status(statusCode.SUCCESS).json({
     code: 0,
     msg: "Create user successful",
```
Comment (kc97ble) [inline controller/auth.js]: `User created`

Try to follow API docs unless there is a special reason.

File: controller/auth.js [hunk 10]
```diff
@@ -17,9 +17,13 @@ function isEmail(email_or_username) {
...
-function sanitizeUserDetails(data) {
+function getHashedPassword(password) {
   const salt = bcrypt.genSaltSync(SALT_ROUNDS);
-  const hashedPassword = bcrypt.hashSync(data.password, salt);
+  return bcrypt.hashSync(password, salt);
+}
+
+function sanitizeUserDetails(data) {
+  const hashedPassword = getHashedPassword(data.password);
   let userDetails = {
```
Comment (kc97ble) [inline controller/auth.js]: use `const` instead of `let`

remove redundant fields

File: controller/user.js [hunk 11]
```diff
@@ -32,7 +35,7 @@ async function getAll(req, res) {
     code: 0,
     msg: "",
     data: {
-      users: users.map(user => buildUserJson(user))
+      users: users.map(user => formatUser(user))
```
Comment (kc97ble) [inline controller/user.js]: rename function to `getAllUsers`

It seems that the function is not needed. Please check the API docs. Comment the function.

File: controller/user.js [hunk 12]
```diff
@@ -47,7 +50,7 @@ async function getByUsername(req, res) {
       code: 0,
       msg: "",
       data: {
-        user: buildUserJson(user)
+        user: formatUser(user)
```
Comment (kc97ble) [inline controller/user.js]: rename function to `getUserByUsername`

set `msg: "User"`, please refer to the API docs

File: controller/user.js [hunk 13]
```diff
@@ -63,7 +66,7 @@ async function getById(req, res) {
...
       },
     });
   }
```
Comment (kc97ble) [inline controller/user.js:35]: `getById` is no longer needed. Refer to API docs. Remove this function.

File: controller/user.js [hunk 14]
```diff
@@ -191,13 +194,43 @@ async function update(req, res, next) {
...
+    if (rowsUpdate == 0) {
+      throw new errors.FcError(errors.SYSTEM_ERROR);
+    }
+    res.status(statusCode.SUCCESS).json({	
+      code: 0,	
+      msg: "Password updated",	
+      data: {},	
+    });
+  })
+  .catch(next);
```
Comment (kc97ble) [inline controller/user.js]:
```
user.password = hashedPassword;
user.save().then(() => res.json{ ... }).catch(next);
```

https://sequelize.org/master/manual/model-instances.html

File: controller/user.js [hunk 15]
```diff
@@ -192,25 +56,34 @@ async function update(req, res, next) {
...
+  if (!bcrypt.compareSync(req.body.old_password, user.password)) {	
+    throw new errors.FcError(errors.EMAIL_USERNAME_PASSWORD_INVALID);	
+  }	
+
+  const hashedPassword = getHashedPassword(req.body.new_password);	
+  user.password = hashedPassword;
+  user.save().then((updatedUser) => {
+    if (!updatedUser) {
+      throw new errors.FcError(errors.SYSTEM_ERROR);
+    }
```
Comment (kc97ble) [inline controller/user.js:72]: remove this unnecessary check

See https://sequelize.org/master/class/lib/model.js~Model.html#instance-method-save

File: controller/user.js [hunk 16]
```diff
@@ -5,8 +5,10 @@ const db = require("../models/index.js");
 const models = db.sequelize.models;
 const jsonwebtoken = require("jsonwebtoken");
 const errors = require("../utils/error");
+const bcrypt = require("bcryptjs");
+const { getHashedPassword } = require("./auth.js");
 
-function buildUserJson(user) {
+function formatUser(user) {
```
Comment (kc97ble) [inline controller/user.js:11]: For `/api/v1/signup`, email is included
For `/api/v1/users/:user_name`, email is not included

Suggestion: change the function signature to:

```
function formatUser(user, publicFieldsOnly=false)
```

When `publicFieldsOnly` is true, exclude `email` from the returned value.

File: controller/user.js (hunk 13 above)
Comment (houtaru) [reply controller/user.js:35]: `getById` is use for GET /api/v1/me

**Pull request merged**